csharp/src/iceboxnet/ServiceManager.cs

```diff
@@ -2,195 +2,125 @@
                 {
-                    throw new NoSuchServiceException();
+                    _ = StartedAsync(observer, startedServices);
                 }
-
-                if (info.Status != ServiceStatus.Stopped)
-                {
-                    throw new AlreadyStartedException();
-                }
-                info.Status = ServiceStatus.Starting;
-                _pendingStatusChanges = true;
+                return default;
             }
 
-            bool started = false;
-            try
+            async Task StartedAsync(IServiceObserverPrx observer, IEnumerable<string> services)
             {
-                Debug.Assert(info.Service != null);
-                Communicator? communicator = info.Communicator ?? _sharedCommunicator;
-                Debug.Assert(communicator != null);
-                info.Service.Start(info.Name, communicator, info.Args);
-                started = true;
-            }
-            catch (Exception ex)
-            {
-                _logger.Warning($"IceBox.ServiceManager: exception while starting service `{info.Name}':\n{ex}");
-            }
-
-            lock (_mutex)
-            {
-                if (started)
+                try
                 {
-                    info.Status = ServiceStatus.Started;
-                    ServiceStarted(name, _observers);
+                    // We don't want to forward the cancel of a dispatch that has already completed.
+                    // The sending / queueing is done with _mutex locked, so barring a connection failure + retry, it
+                    // will reach the observer before subsequent state change notifications.
+                    await observer.ServicesStartedAsync(services, cancel: default);
                 }
-                else
+                catch (Exception ex)
                 {
-                    info.Status = ServiceStatus.Stopped;
+                    RemoveObserver(observer, ex);
                 }
-                _pendingStatusChanges = false;
-                Monitor.PulseAll(_mutex);
             }
         }
 
-        public void StopService(string name, Current current, CancellationToken cancel)
+        public ValueTask ShutdownAsync(Current current, CancellationToken cancel)
         {
-            ServiceInfo? info;
-            lock (_mutex)
-            {
-                // Search would be more efficient if services were contained in a map, but order is required for
-                // shutdown.
-                info = _services.Find(service => service.Name.Equals(name));
-                if (info == null)
-                {
-                    throw new NoSuchServiceException();
-                }
-
-                if (info.Status != ServiceStatus.Started)
-                {
-                    throw new AlreadyStoppedException();
-                }
-                _pendingStatusChanges = true;
-            }
+            _ = _communicator.ShutdownAsync();
```

Comment:
We'd have to keep around the cancellation token source here as well to be able to cancel it... so perhaps not such a good idea.